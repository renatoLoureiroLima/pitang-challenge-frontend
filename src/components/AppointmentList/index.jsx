import { format } from 'date-fns';
import { useEffect, useState } from 'react';
import { Table } from 'react-bootstrap';
import { toast } from 'react-toastify';
import axios from '../../services/api';

function AppointmentList() {
  const [appointments, setAppointments] = useState(null);

  const fetchData = async () => {
    await axios
      .get('/appointment')
      .then((response) => {
        setAppointments(
          response.data.appointments.map((appointment) => {
            return {
              ...appointment,
              birthday: new Date(appointment.birthday),
              selectedDate: new Date(appointment.selectedDate),
            };
          })
        );
      })
      .catch(() => toast.error('Oops, aconteceu algo de errado!'));
  };

  const dateMask = (date) => {
    const dashed = `${date.getDate()}/${
      date.getMonth() + 1
    }/${date.getFullYear()}`;

    return dashed;
  };

  const dateMaskHour = (date) => {
    const dashed = format(new Date(date), "dd/MM/yyyy' - 'HH:mm");

    return dashed;
  };

  const isAttended = async ({ target: { checked } }, appointment) => {
    const newAppointments = appointments.map((newAppointment) => {
      if (newAppointment._id === appointment._id) {
        return {
          ...newAppointment,
          attended: checked,
        };
      }
      return newAppointment;
    });

    await axios
      .put(`/appointment/${appointment._id}`, {
        name: appointment.name,
        attended: checked,
      })
      .then(
        checked
          ? toast.success('Atendimento concluído')
          : toast.error('Atendimento não concluído')
      );

    setAppointments(newAppointments);
  };

  useEffect(() => {
    if (!appointments) fetchData();
  }, []);

  return (
    <>
      <Table data-testid='appointmentlist-test'>
        <thead>
          <tr>
            <th>Nome: </th>
            <th>Data de nascimento </th>
            <th>Data do Atendimento </th>
            <th>Antedimento </th>
          </tr>
        </thead>
        <tbody>
          {appointments &&
            appointments
              .sort((a, b) => {
                return new Date(a.selectedDate) - new Date(b.selectedDate);
              })
              .map((appointment) => (
                <tr key={appointment.index}>
                  <td>{appointment.name}</td>
                  <td>{dateMask(appointment.birthday)}</td>
                  <td>{dateMaskHour(appointment.selectedDate)}</td>
                  <td>
                    <input
                      onChange={(event) => isAttended(event, appointment)}
                      checked={appointment.attended}
                      type='checkbox'
                    />

                    <span className='m-4'>
                      {appointment.attended
                        ? 'Concluído'
                        : 'Aguardando atendimento'}
                    </span>
                  </td>
                </tr>
              ))}
        </tbody>
      </Table>
    </>
  );
}

export default AppointmentList;
