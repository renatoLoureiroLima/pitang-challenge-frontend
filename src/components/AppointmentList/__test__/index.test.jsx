import ReactDOM from 'react-dom';
import AppointmentList from '../index';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AppointmentList></AppointmentList>, div);
});

it('renders table properly', () => {
  const { getByTestId } = render(<AppointmentList></AppointmentList>);
  expect(getByTestId('appointmentlist-test')).toHaveDisplayValue;
});

it('matches snapshot', () => {
  const tree = renderer.create(<AppointmentList></AppointmentList>).toJSON();
  expect(tree).toMatchSnapshot();
});
