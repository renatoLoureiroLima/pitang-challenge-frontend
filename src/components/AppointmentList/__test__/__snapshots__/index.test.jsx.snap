// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<table
  className="table"
  data-testid="appointmentlist-test"
>
  <thead>
    <tr>
      <th>
        Nome: 
      </th>
      <th>
        Data de nascimento 
      </th>
      <th>
        Data do Atendimento 
      </th>
      <th>
        Antedimento 
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
